lib/src/rules/no_implicit_dynamic.dart

```diff
@@ -0,0 +1,202 @@
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.annotate_types;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/linter.dart';
+import 'package:linter/src/util.dart';
+
+const desc = 'Implicit use of dynamic.';
+
+const details = '''
+**AVOID** using "implicitly dynamic" values.
+
+Untyped / dynamic invocations may fail or be slower at runtime, but dynamic
+types often creep up unintentionally. Explicitly mark variables or return types
+as `dynamic` (instead of `var`) to express your intent unequivocally.
+
+Note: this works best with the --strong command-line flag and after disabling
+both `always_specify_types` and `always_declare_return_types` lints.
+
+**GOOD:**
+```dart
+String trim(String s) => s.trim();
+
+main() {
+  var s = trim(' a ').toUpperCase();
+
+  dynamic x;
+  x = ... ;
+  x.reallyNotSureThisExists();
+}
+```
+
+**BAD:**
+```dart
+trim(s) => s.trim();
+
+main() {
+  var s = trim(1).toUpperCase();
+
+  var x;
+  x = ... ;
+  x.reallyNotSureThisExists();
+}
+```
+''';
+
+class NoImplicitDynamic extends LintRule {
+  NoImplicitDynamic()
+      : super(
+            name: 'no_implicit_dynamic',
+            description: desc,
+            details: details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+// TODO(ochafik): Handle implicit return types of method declarations (vs. overrides).
+class Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  Visitor(this.rule);
+
+  Element _getBestElement(Expression node) {
+    if (node is SimpleIdentifier) return node.bestElement;
+    if (node is PrefixedIdentifier) return node.bestElement;
+    if (node is PropertyAccess) return node.propertyName.bestElement;
+    return null;
+  }
+
+  bool _isImplicitDynamic(Expression node) {
+    if (node == null) return false;
+    while (node is ParenthesizedExpression) {
+      node = node.expression;
+    }
+
+    if (node is AsExpression || node is Literal) return false;
+    var t = node.bestType;
+    if (!t.isDynamic && !t.isObject) return false;
+
+    var e = _getBestElement(node);
+    if (e is PropertyAccessorElement) e = e.variable;
+    if (e is VariableElement) return e.hasImplicitType;
+
+    if (node is ConditionalExpression) {
+      return !node.thenExpression.bestType.isDynamic ||
+         !node.elseExpression.bestType.isDynamic;
+    }
+    if (node is MethodInvocation) {
+      return node.methodName.bestElement?.hasImplicitReturnType != false;
+    }
+
+    return true;
+  }
+
+  void _checkTarget(Expression target, [token]) {
+    if (_isImplicitDynamic(target)) {
+      // Avoid double taxation (if `x` is dynamic, only lint `x.y.z` once).
+      Expression subTarget;
+      if (target is PropertyAccess) subTarget = target.realTarget;
```

Comment:
Perhaps it would be good idea to create a method that return an iterable with all targets, for example when you have something like:

```dart
a.b.foo().c.d.bar()[3].baz();
```

So if you can create a method that could be use in unnecessary_lambdas too, where I had to do a similar check